### PR TITLE
Heatmap add cell sets check

### PR DESCRIPTION
**Pull request by aerlaut**
# Background
#### Link to issue 

https://this-is-biomage.slack.com/archives/C0250TDU1PF/p1629886977002000

#### Link to staging deployment URL 

Coming soon

#### Links to any Pull Requests related to this

N.A.

#### Anything else the reviewers should know about the changes here

Add checks for empty cell set

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [X] Unit tests written
- [X] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [X] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: src/__test__/components/data-exploration/heatmap/HeatmapPlot.test.jsx [hunk 1]
```diff
@@ -252,6 +256,32 @@ describe('HeatmapPlot', () => {
...
+
+    component = mount(
+      <Provider store={store}>
+        <HeatmapPlot experimentId={experimentId} width={200} height={200} />
+      </Provider>,
+    );
+
+    expect(component.find('VegaHeatmap').length).toEqual(0);
+
+    console.log(component.debug());
```
Comment (ivababukova) [inline src/__test__/components/data-exploration/heatmap/HeatmapPlot.test.jsx]: remove this console log

File: src/__test__/components/data-exploration/heatmap/HeatmapPlot.test.jsx [hunk 2]
```diff
@@ -252,6 +256,30 @@ describe('HeatmapPlot', () => {
...
+    component = mount(
+      <Provider store={store}>
+        <HeatmapPlot experimentId={experimentId} width={200} height={200} />
+      </Provider>,
+    );
+
+    expect(component.find('VegaHeatmap').length).toEqual(0);
+
+    // 2 Because the Loader component contains ClipLoader, which contains a Loader
+    expect(component.find('Loader').length).toEqual(2);
```
Comment (ivababukova) [inline src/__test__/components/data-exploration/heatmap/HeatmapPlot.test.jsx]: Agi, if you comment out the changes you did in `src/components/data-exploration/heatmap/HeatmapPlot.jsx`, would this test fail? Just to check if this test is catching this bug

Comment (aerlaut) [reply src/__test__/components/data-exploration/heatmap/HeatmapPlot.test.jsx]: yes, the test fails

File: src/components/data-exploration/heatmap/HeatmapPlot.jsx [hunk 3]
```diff
@@ -193,7 +188,11 @@ const HeatmapPlot = (props) => {
     );
   }
 
-  if (isHeatmapGenesLoading) {
+  if (
+    isHeatmapGenesLoading
+    || cellSetsLoading
+    || cellSetsHierarchy.length === 0
```
Comment (ivababukova) [inline src/components/data-exploration/heatmap/HeatmapPlot.jsx]: this doesn't sound right: if `cellSetsHierarchy` is 0, then we shouldn't be in loading state, then we will just have an empty component, because there will be no cell sets to display gene expression for

File: src/__test__/components/data-exploration/heatmap/HeatmapPlot.test.jsx (hunk 2 above)
Comment (aerlaut) [reply src/__test__/components/data-exploration/heatmap/HeatmapPlot.test.jsx]: It renders the heatmap instead of the loader

Comment (ivababukova) [reply src/__test__/components/data-exploration/heatmap/HeatmapPlot.test.jsx]: ok, great.

File: src/__test__/components/data-exploration/heatmap/HeatmapPlot.test.jsx [hunk 4]
```diff
@@ -252,6 +256,30 @@ describe('HeatmapPlot', () => {
...
+
+    component = mount(
+      <Provider store={store}>
+        <HeatmapPlot experimentId={experimentId} width={200} height={200} />
+      </Provider>,
+    );
+
+    expect(component.find('VegaHeatmap').length).toEqual(0);
+
+    // 2 Because the Loader component contains ClipLoader, which contains a Loader
```
Comment (ivababukova) [inline src/__test__/components/data-exploration/heatmap/HeatmapPlot.test.jsx]: just noticed and commented in the code below: if cellSets are empty, we should be displaying an empty component, not loading component. Make sure you fix this unit test too.

File: src/components/data-exploration/heatmap/HeatmapPlot.jsx (hunk 3 above)
Comment (aerlaut) [reply src/components/data-exploration/heatmap/HeatmapPlot.jsx]: Oh, I see what you mean. I thought we're handling the case where cellSets is empty while waiting for the cell sets to load. So this is to handle the case where cell sets is empty, even after cell sets are loaded?

Comment (ivababukova) [reply src/components/data-exploration/heatmap/HeatmapPlot.jsx]: I don't understand this comment Agi. I don't see how `cellSetsHierarchy` should have any impact on the decision whether the heatmap should be in a loading state or not. Wouldn't just `cellSetsLoading` be enough on the cell sets side?

Comment (ivababukova) [reply src/components/data-exploration/heatmap/HeatmapPlot.jsx]: maybe we should have a short chat about this, maybe we are talking about different things here

File: src/__test__/components/data-exploration/heatmap/HeatmapPlot.test.jsx (hunk 4 above)
Comment (aerlaut) [reply src/__test__/components/data-exploration/heatmap/HeatmapPlot.test.jsx]: Fixed

File: src/__test__/components/data-exploration/heatmap/HeatmapPlot.test.jsx (hunk 1 above)
Comment (aerlaut) [reply src/__test__/components/data-exploration/heatmap/HeatmapPlot.test.jsx]: removed

File: src/components/data-exploration/heatmap/HeatmapPlot.jsx [hunk 5]
```diff
@@ -223,6 +227,10 @@ const HeatmapPlot = (props) => {
     );
   }
 
+  const signalListeners = {
+    mouseOver: handleMouseOver,
+  };
+
```
Comment (ivababukova) [inline src/components/data-exploration/heatmap/HeatmapPlot.jsx:233]: why is that added?

Comment (ivababukova) [reply src/components/data-exploration/heatmap/HeatmapPlot.jsx:233]: nevermind

**Review by ivababukova — APPROVED** [review]

File: src/components/data-exploration/heatmap/HeatmapPlot.jsx (hunk 5 above)
Comment (aerlaut) [reply src/components/data-exploration/heatmap/HeatmapPlot.jsx:233]: it's not added, just moved
